### PR TITLE
Add code block copy button

### PR DESCRIPTION
**Pull request by compulim**
> Fixes #

### Added

-  Added copy button to code blocks, in PR [#5334](https://github.com/microsoft/BotFramework-WebChat/pull/5334), by [@compulim](https://github.com/compulim)

## Description

For code blocks inside Markdown (via triple-backtick or `<pre>`), we will add a copy button on the top right corner of it.

image

## Design

We are introducing internal-use Web Components.

As we go through sanitization, DOM become string and it lose event handler. In other words, we cannot listen to "click" event on the button after sanitization.

Our solution is to use Web Components so we can preserve logic during sanitization.

Alternatively, we could have a catch-all click handler on Markdown, but it is not trivial as Adaptive Cards requires a HTML string instead of DOM too. That means, we will need to put a catch-all container in 2+ different places.

## Specific Changes

- Added new `CustomElementsProvider` to host internal-use hash-secured Web Components
- Added new code block copy button `<webchat-xxxxx--code-block-copy-button>` where `xxxxx` is a random hash
- Chaining `package.json` via `touch` so changes to `core` will trigger rebuild of `api`, and so on

-  [x] I have added tests and executed them locally
-  [x] I have updated `CHANGELOG.md`
-  [x] ~I have updated documentation~

## Review Checklist

> This section is for contributors to review your work.

-  [x] Accessibility reviewed (tab order, content readability, alt text, color contrast)
-  [x] ~Browser and platform compatibilities reviewed~
-  [x] CSS styles reviewed (minimal rules, no `z-index`)
-  [x] ~Documents reviewed (docs, samples, live demo)~
-  [x] ~Internationalization reviewed (strings, unit formatting)~
-  [x] ~`package.json` and `package-lock.json` reviewed~
-  [x] Security reviewed (no data URIs, check for nonce leak)
-  [x] Tests reviewed (coverage, legitimacy)

### CODE REVIEW

**Review by OEvgeny — DISMISSED** [review]

File: packages/component/src/providers/CustomElements/customElements/registerCodeBlockCopyButton.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+      );
+
+      const buttonElement = this.ownerDocument.createElement('button');
+
+      buttonElement.classList.add('webchat__code-block-copy-button', styleSet.codeBlockCopyButton);
+      buttonElement.dataset.testid = testIds.codeBlockCopyButton;
+      buttonElement.type = 'button';
+
+      buttonElement.append(copyIconImageElement);
+      buttonElement.append(copiedIconImageElement);
```
Comment (OEvgeny) [inline packages/component/src/providers/CustomElements/customElements/registerCodeBlockCopyButton.ts:40]: Using html module would be a nice touch here I guess.

**Review by OEvgeny — APPROVED** [review]

**Review by OEvgeny — APPROVED** [review]

File: packages/component/src/providers/CustomElements/customElements/registerCodeBlockCopyButton.ts [hunk 2]
```diff
@@ -9,7 +9,8 @@ type Strings = Readonly<{
 type StyleSet = ReturnType<typeof useStyleSet>[0];
 
 export default function registerCodeBlockCopyButton(hash: string, strings: Strings, styleSet: StyleSet): string {
-  const name = `webchat-${hash}__code-block-copy-button`;
+  // Allowed tag names are ASCII alphanumeric only, https://html.spec.whatwg.org/multipage/syntax.html#tag-name.
+  const name = `WebChatCodeBlockCopyButton${hash}`;
```
Comment (OEvgeny) [inline packages/component/src/providers/CustomElements/customElements/registerCodeBlockCopyButton.ts]: I think it should have at least one hyphen, and AFAIK html names are case insensitive

File: packages/component/src/providers/CustomElements/customElements/registerCodeBlockCopyButton.ts [hunk 3]
```diff
@@ -1,12 +1,8 @@
...
 
-type ObservedAttributes = 'className' | 'data-alt-copied' | 'data-alt-copy';
+type ObservedAttributes = 'class' | 'data-alt-copied' | 'data-alt-copy';
 
-const observedAttributes: readonly ObservedAttributes[] = Object.freeze([
-  'className',
-  'data-alt-copied',
-  'data-alt-copy'
-]);
+const observedAttributes: readonly ObservedAttributes[] = Object.freeze(['class', 'data-alt-copied', 'data-alt-copy']);
```
Comment (OEvgeny) [inline packages/component/src/providers/CustomElements/customElements/registerCodeBlockCopyButton.ts:5]: Welcome back to html world :rofl:

**Review by OEvgeny — DISMISSED** [review]

File: packages/component/src/providers/CustomElements/customElements/registerCodeBlockCopyButton.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+export default function registerCodeBlockCopyButton(hash: string): string {
+  // Allowed tag names are specified here, https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define#valid_custom_element_names
+  const name = `webchat-${hash}--code-block-copy-button`;
+
+  customElements.define(
+    name,
+    // Using anonymous class because one class can only be registered with one tag name.
+    // Web Components enforce 1:1 relationship.
+    class extends CodeBlockCopyButtonElement {
```
Comment (OEvgeny) [inline packages/component/src/providers/CustomElements/customElements/registerCodeBlockCopyButton.ts:117]: If you don't want to create a chain of inheritance, it should be possible to wrap the class creation into a function and call here.

On the other hand this approach allows ckecking if we derived from the initial class.

Overall, there shouldn't be a problem keeping 1-1 relationship between Web Chat instances using React primitives. I assume something else needs our attention.

**Review by OEvgeny — DISMISSED** [review]

**Review by OEvgeny — APPROVED** [review]

File: packages/component/src/providers/CustomElements/customElements/registerCodeBlockCopyButton.ts (hunk 2 above)
Comment (compulim) [reply packages/component/src/providers/CustomElements/customElements/registerCodeBlockCopyButton.ts]: All lowercase, at least one hyphen. :)

**Pull request merged**